[ZIP 230, ZIP 246] v6 transaction format and sighash (#987)
zcash/zips: zips/zip-0230.rst

+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 4                           |``nExpiryHeight``             |``uint32``                                      |A block height in the range {1 .. 499999999} after which             |
|                             |                              |                                                |the transaction will expire, or 0 to disable expiry. [#zip-0203]_    |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 8                           |``fee``                       |``uint64``                                      |The fee to be paid by this transaction, in zatoshis.                 |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | 8                           |``burnAmount``                |``uint64``                                      |The value to be burned in this transaction, in zatoshis.             |

Reviewer: nit: Is this field name changing, or are we still calling it "burning"?

> | 8                           |``burnAmount``                |``uint64``                                      |The value to be burned in this transaction, in zatoshis.             |

Reviewer: ```suggestion
| 8                           |``zip233Amount``                |``uint64``                                      |The value to be burned in this transaction, in zatoshis.             |
```

Reviewer: This change can be made in a future PR, to avoid foramtting issues.

| varies                      |``tx_out``                    |``tx_out``                                      |Transparent outputs, encoded as in Bitcoin.                          |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| **Sapling Transaction Fields**                                                                                                                                                    |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| varies                      |``nSpendsSapling``            |``compactSize``                                 |Number of Sapling Spend descriptions in ``vSpendsSapling``.          |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | 96 × ``nSpendsSapling``     |``vSpendsSapling``            |``SpendDescriptionV6[nSpendsSapling]``          |A sequence of Sapling Spend descriptions, encoded per                |
> |                             |                              |                                                |protocol §7.3 ‘Spend Description Encoding and Consensus’.            |

Reviewer: ```suggestion
| 96 × ``nSpendsSapling``     |``vSpendsSapling``            |``SpendDescriptionV5[nSpendsSapling]``          |A sequence of Sapling Spend descriptions, encoded per                |
|                             |                              |                                                |protocol §7.3 ‘Spend Description Encoding and Consensus’  (unchanged from V5). |
```

+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| varies                      |``nOutputsSapling``           |``compactSize``                                 |Number of Sapling Output Decriptions in ``vOutputsSapling``.         |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | 756 × ``nOutputsSapling``   |``vOutputsSapling``           |``OutputDescriptionV6[nOutputsSapling]``        |A sequence of Sapling Output descriptions, encoded per               |

Reviewer: ```suggestion
| 276 × ``nOutputsSapling``   |``vOutputsSapling``           |``OutputDescriptionV6[nOutputsSapling]``        |A sequence of Sapling Output descriptions, encoded per               |
```

|                             |                              |                                                |`OrchardZSA Action Group Description`_.                              |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 8                           |``valueBalanceOrchard``       |``int64``                                       |The net value of Orchard spends minus outputs.                       |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 64                          |``bindingSigOrchard``         |``byte[64]``                                    |An OrchardZSA binding signature on the SIGHASH transaction hash.     |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | **OrchardZSA Issuance Fields**                                                                                                                                                    |

Reviewer: ```suggestion
| **ZSA Issuance Fields**                                                                                                                                                    |
```

+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| varies                      |``nIssueActions``             |``compactSize``                                 |The number of issuance actions in the bundle.                        |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | ``IssueActionSize`` ×       |``IssueAction[nIssueActions]``|                                                |A sequence of issuance action descriptions, where IssueActionSize is |
> | ``nIssueActions``           |                              |                                                |the size, in bytes, of an IssueAction description.                   |

Reviewer: ```suggestion
| varies                      |``IssueAction[nIssueActions]``|                                                |A sequence of issuance action descriptions, where IssueActionSize is |
|                             |                              |                                                |the size, in bytes, of an IssueAction description.                   |
```

| 8                           |``valueBalanceOrchard``       |``int64``                                       |The net value of Orchard spends minus outputs.                       |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| varies                      | ``nAssetBurn``               | ``compactSize``                                |The number of Assets burnt.                                          |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 40 × ``nAssetBurn``         | ``vAssetBurn``               | ``AssetBurn[nAssetBurn]``                      |A sequence of Asset Burn descriptions, encoded per                   |
|                             |                              |                                                | encoded per `OrchardZSA Asset Burn Description`_.                   |

> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

Author: As noted in https://github.com/zcash/zips/pull/960#issuecomment-2635241787, this change requires further discussion. In particular, no rationale has been provided for this move (which should be in ZIP 227), and there's been no discussion anywhere I'm aware of about the privacy implications of this change, or how it impacts the ability of the ActionGroup bundler to decide which ActionGroups to bundle (AFAICT after this change, the bundler would need to contribute their own ActionGroup in order to handle any mismatch that requires altering burns, which increases both the size and fee of the transaction beyond what the bundled ActionGroups would otherwise cost).

> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

Reviewer 2: - For the NU7 setting, there is no impact as such since the bundling continues to be done by the party creating the component action groups. 
- For the subsequent Swaps addition, we already expect the bundler to create their own Action Group in order to collect any "matching fees", so this is not a cost that is not already part of the Swaps mechanism.

> any mismatch that requires altering burns

An example is when someone creates an Action Group with burn fields corresponding to Assets they don’t have. If the bundler wants to accept such an Action Group by adding the necessary Assets to the transaction themselves, they are free to do so (though I don't see why they would?). If not, they are free to ignore the Action Group that doesn’t balance out, so such Action Groups would never be included. Did you have a different scenario in mind? 

The motivation for having burn information inside the Action Group is to allow parties to be able to burn Assets as well, otherwise the burning ability remains only with the bundler. (I agree this rationale should be added)

> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

Reviewer 2: Also, I suppose it is better to have this discussion in its own PR rather than here?

We already have this change up in #976 building on top of main, though it does not have the ZIP 246 changes which are part of this PR. I also opened #991 that builds on top of this PR (#987) instead.

> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

Author: Indeed, these changes should have initially been proposed in their own PR (#976 was not such a PR, as it also included other unrelated changes). The discussion continues here: https://github.com/zcash/zips/pull/991/files#r1990147143

> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

Author: PR has now been merged into this one.

> 

Reviewer 3: At line 191, change ``SpendDescriptionV6`` to ``SpendDescriptionV5``, and add an extra bullet point:
```rst
* There is no ``SpendDescriptionV6`` because the format of Sapling Spend descriptions
  has not changed from SpendDescriptionV5`` [#zip-0225]_.
```

* If ``fAllPruned == 0``, then:

  * ``nonceOrHash`` contains the ZIP 231 nonce for deriving encryption keys.

  * Each bit of ``pruned``, in little-endian order, indicates the type of the corresponding
    entry in ``vMemoChunks``.

    * A bit value of 0 indicates that the entry will be of type ``byte[272]`` representing an
      encrypted memo chunk.

    * A bit value of 1 indicates the entry will be a ``byte[32]`` and contains the ZIP 246
      ``memo_chunk_digest`` for a pruned chunk.

* If ``fAllPruned == 1``, then:

  * ``nonceOrHash`` contains the ZIP 246 ``memo_digest`` for the pruned memo bundle.

The encodings of ``tx_in``, and ``tx_out`` are as in a version 4 transaction (i.e.
unchanged from Canopy). The encodings of ``SpendDescriptionV6``, ``OutputDescriptionV6``
, ``ActionGroupDescription``, ``AssetBurn`` and ``IssueAction`` are described below.
The encoding of Sapling Spends and Outputs has changed relative to prior versions in order
to better separate data that describe the effects of the transaction from the proofs of and
commitments to those effects, and for symmetry with this separation in the Orchard-related parts
of the transaction format.

> Sapling Spend Description (``SpendDescriptionV6``)
> --------------------------------------------------
> 

Reviewer: ```suggestion
```

> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
> | Bytes                       | Name                         | Data Type                                      | Description                                                         |
> +=============================+==============================+================================================+=====================================================================+
> | 32                          |``cv``                        |``byte[32]``                                    |A value commitment to the net value of the input note.               |
> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
> | 32                          |``nullifier``                 |``byte[32]``                                    |The nullifier of the input note.                                     |
> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
> | 32                          |``rk``                        |``byte[32]``                                    |The randomized validating key for the element of                     |
> |                             |                              |                                                |``spendAuthSigsSapling`` corresponding to this Spend.                |
> +-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
> 
> The encodings of each of these elements are defined in §7.3 ‘Spend Description Encoding
> and Consensus’ of the Zcash Protocol Specification [#protocol-spenddesc]_.
> 

Reviewer: ```suggestion
```

+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 32                          |``cmu``                       |``byte[32]``                                    |The :math:`u\!`-coordinate of the note commitment for the output     |
|                             |                              |                                                |note.                                                                |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 32                          |``ephemeralKey``              |``byte[32]``                                    |An encoding of an ephemeral Jubjub public key.                       |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | 580                         |``encCiphertext``             |``byte[580]``                                   |The encrypted contents of the note plaintext.                        |

Reviewer: ```suggestion
| 100                         |``encCiphertext``             |``byte[580]``                                   |The encrypted contents of the note plaintext.                        |
```

+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| Bytes                       | Name                         | Data Type                                      | Description                                                         |
+=============================+==============================+================================================+=====================================================================+
| varies                      |``nActionsOrchard``           |``compactSize``                                 |The number of Action descriptions in ``vActionsOrchard``.            |
|                             |                              |                                                |This MUST have a value strictly greater than ``0``.                  |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | 852 × ``nActionsOrchard``   |``vActionsOrchard``           |``OrchardZSAAction[nActionsOrchard]``           |A sequence of OrchardZSA Action descriptions in the Action Group.    |

Reviewer: ```suggestion
| 372 × ``nActionsOrchard``   |``vActionsOrchard``           |``OrchardZSAAction[nActionsOrchard]``           |A sequence of OrchardZSA Action descriptions in the Action Group.    |
```

+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 1                           |``flagsOrchard``              |``byte``                                        |As defined in Section 7.1 of the Protocol Specification              |

> |                             |                              |                                                |[#protocol-txnencoding]_.                                            |

Reviewer: These should reference ZIP 225, or a more precise reference within the protocol spec, instead of section 7.1.

  ``vSpendAuthSigsOrchard``, each have a 1:1 correspondence to the elements of
  ``vActionsOrchard`` and MUST be ordered such that the proof or signature at a given
  index corresponds to the ``OrchardZSAAction`` at the same index.

> * If thee value of ``nAGExpiryHeight`` is nonzero, it MUST be consistent with

Reviewer 4: ```suggestion
* If the value of ``nAGExpiryHeight`` is nonzero, it MUST be consistent with
```

> For the OrchardZSA protocol, which does not make use of an additional expiry height for transactions, we set the value of ``nAGExpiryHeight`` to be ``0`` by consensus.

Reviewer: We need to ensure that a conformance requirement exists within the specification that requires `nAGExpiryHeight` to be zero as of NU7.

+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 32                          |``cmx``                       |``byte[32]``                                    |The :math:`x\!`-coordinate of the note commitment for the output     |
|                             |                              |                                                |note.                                                                |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| 32                          |``ephemeralKey``              |``byte[32]``                                    |An encoding of an ephemeral Pallas public key.                       |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | 612                         |``encCiphertext``             |``byte[580]``                                   |The encrypted contents of the note plaintext.                        |

Reviewer: ```suggestion
| 132                         |``encCiphertext``             |``byte[132]``                                   |The encrypted contents of the note plaintext.                        |
```

An OrchardZSA Asset Burn description is encoded in a transaction as an instance of an ``AssetBurn`` type:

+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+
| Bytes                       | Name                         | Data Type                                      | Description                                                         |
+=============================+==============================+================================================+=====================================================================+
| 32                          |``AssetBase``                 |``byte[32]``                                    |For the OrchardZSA protocol, this is the encoding of the Asset Base  |
|                             |                              |                                                |:math:`\mathsf{AssetBase^{Orchard}}\!`.                              |
+-----------------------------+------------------------------+------------------------------------------------+---------------------------------------------------------------------+

> | 8                           |``valueBurn``                 |``uint64``                                      |The amount being burnt. The value is checked by consensus to be      |

Reviewer: Given that this *is* a burn, in contrast to the ZSF deposit, I think it's important to have different naming.
